tempesta_fw/apm.c

```diff
@@ -1151,14 +1496,318 @@ tfw_cfgop_apm_stats(TfwCfgSpec *cs, TfwCfgEntry *ce)
 	return 0;
 }
 
+static int
+tfw_cfgop_apm_server_failover(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	TfwApmHMCfg *hm_entry;
+	int code, limit, tframe;
+
+	if (tfw_cfg_check_val_n(ce, 3))
+		return -EINVAL;
+
+	if (ce->attr_n) {
+		TFW_ERR_NL("Unexpected attributes\n");
+		return -EINVAL;
+	}
+
+	if (tfw_cfgop_parse_http_status(ce->vals[0], &code)) {
+		TFW_ERR_NL("Unable to parse http code value: '%s'\n",
+			   ce->vals[0]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_parse_int(ce->vals[1], &limit)) {
+		TFW_ERR_NL("Unable to parse http limit value: '%s'\n",
+			   ce->vals[1]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_check_range(limit, 1, USHRT_MAX))
+		return -EINVAL;
+
+	if (tfw_cfg_parse_int(ce->vals[2], &tframe)) {
+		TFW_ERR_NL("Unable to parse http tframe value: '%s'\n",
+			   ce->vals[2]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_check_range(tframe, 1, USHRT_MAX))
+		return -EINVAL;
+
+	hm_entry = kzalloc(sizeof(TfwApmHMCfg), GFP_KERNEL);
+	if (!hm_entry)
+		return -ENOMEM;
+
+	INIT_LIST_HEAD(&hm_entry->list);
+	hm_entry->code = code;
+	hm_entry->limit = limit;
+	hm_entry->tframe = tframe;
+	list_add(&hm_entry->list, &tfw_hm_codes_list);
+	++tfw_hm_codes_cnt;
+
+	return 0;
+}
+
+static void
+tfw_cfgop_apm_cleanup_server_failover(TfwCfgSpec *cs)
+{
+	TfwApmHMCfg *ent, *tmp;
+
+	list_for_each_entry_safe(ent, tmp, &tfw_hm_codes_list, list) {
+		list_del_init(&ent->list);
+		kfree(ent);
+	}
+	INIT_LIST_HEAD(&tfw_hm_codes_list);
+}
+
+static int
+tfw_cfgop_begin_apm_hm(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	TfwApmHM *hm;
+	size_t size;
+
+	if (tfw_cfg_check_val_n(ce, 1))
+		return -EINVAL;
+
```

Comment:
My understanding is that these two checks are congenerical -  about a very similar thing, simple, and logically in the same realm. Usually there's no need for an empty line between checks like these as they are not logically separated. There's lot of examples in the code for this.

Reply:
Corrected (and two cases below too).